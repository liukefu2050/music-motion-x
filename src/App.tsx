/* eslint-disable react/require-default-props */
/* eslint-disable react/forbid-prop-types */

import React, { Suspense } from "react"
import { StaticRouter, StaticRouterProps } from "react-router-dom/server"
import {
  QueryClient,
  QueryClientProvider,
  useQueryErrorResetBoundary,
} from "react-query"
import { BrowserRouter } from "react-router-dom"
import { ErrorBoundary } from "react-error-boundary"
import { Provider } from "react-redux"
import { RouteMatch, Routes, Route } from "react-router"
import ErrorFound from "@/components/ErrorPage"

import routes from "./routes"
import NotFound from "./pages/NotFound"
import PlayBar from "@/components/PlayBar"
import AppUpdateAvailable from "@/components/AppUpdateAvailable"

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      suspense: true,
    },
  },
})

const App = ({
  store,
  isServer,
  location,
  preloadedState,
}: {
  store: any
  isServer: boolean
  matchedRoutes?: RouteMatch[]
  location?: StaticRouterProps["location"]
  preloadedState: { [x: string]: any }
}) => {
  const { reset } = useQueryErrorResetBoundary()
  const content = (
    <>
      {/* <PageBack isBlack={false} /> */}
      <Suspense>
        <PlayBar />
      </Suspense>

      <Routes>
        {routes.map((r) => (
          <Route element={r.element} path={r.path} key={r.path}>
            {r.children &&
              r.children.length > 0 &&
              r.children.map((child) => (
                <Route
                  element={child.element}
                  path={child.path}
                  key={child.path}
                />
              ))}
          </Route>
        ))}
        <Route path="*" element={<NotFound />} />
      </Routes>
    </>
  )

  const IsomophicRouter = isServer ? (
    <StaticRouter location={location || ""}>{content}</StaticRouter>
  ) : (
    <BrowserRouter>
      <>
        <AppUpdateAvailable />
        {content}
      </>
    </BrowserRouter>
  )

  return (
    <ErrorBoundary
      onReset={reset}
      fallbackRender={({ resetErrorBoundary, error }) => (
        <ErrorFound resetErrorBoundary={resetErrorBoundary} error={error} />
      )}
    >
      <Provider store={store} serverState={preloadedState || {}}>
        <QueryClientProvider client={queryClient}>
          {IsomophicRouter}
        </QueryClientProvider>
      </Provider>
    </ErrorBoundary>
  )
}

export default App
