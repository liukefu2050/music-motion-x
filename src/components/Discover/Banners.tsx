import { useQuery } from "react-query"
import cx from "classnames"
import fetcher from "@/fetcher"
import type Banner from "@/interfaces/banner"
import { memo, useCallback, useRef, useState } from "react"
import MyBanner from "./MyBanner"

const Banners = memo(() => {
  const [activeBanner, setActiveBanner] = useState(0)

  const myBannerref = useRef<{ nextPic: (idx: number) => void }>({
    nextPic: (_id) => {},
  })

  const onBannerChange = useCallback((index: number) => {
    setActiveBanner(index)
  }, [])

  const { data } = useQuery("/api/banner?type=2", () =>
    fetcher
      .get<{ banners: Banner[] }>("/api/banner?type=2")
      .then((res) => res.data.banners)
      .then((banners) =>
        banners.map((banner) => ({
          pic: banner.pic,
        })),
      ),
  )
  return (
    <div
      className=" w-full relative overflow-hidden"
      style={{ padding: "39% 0 0" }}
    >
      <div className=" absolute bottom-1 left-1/2 -translate-x-1/2 z-20 will-change-transform">
        {data?.map((_, idx) => (
          <li
            className={cx(" mr-1 w-[6px] h-[6px] inline-block rounded-[50%]", {
              "bg-secondary": activeBanner === idx,
              " bg-white": activeBanner !== idx,
            })}
            key={idx}
            onClick={() => {
              myBannerref.current.nextPic(idx)
            }}
          />
        ))}
      </div>
      <MyBanner
        banners={
          data?.map((b) => (b.pic ? b.pic.replace(/https?/, "https") : "")) ??
          []
        }
        onBannerChange={onBannerChange}
        ref={myBannerref}
      />
    </div>
  )
})

Banners.displayName = "Banners"

export default Banners
