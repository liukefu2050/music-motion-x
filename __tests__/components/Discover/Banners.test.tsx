import Banners from "../../../src/components/Discover/Banners"
import nock from "nock"
import App, { queryClient } from "../../test-utils/StoreProvider"
import { render, waitFor, screen, fireEvent } from "@testing-library/react"

describe("Banners", () => {
  beforeAll(() => {
    nock.disableNetConnect()
  })

  afterEach(() => {
    nock.cleanAll()
  })
  afterAll(() => {
    nock.enableNetConnect()
  })

  it("render well without images", async () => {
    nock("https://81.69.200.140").get(`/api/banner?type=2`).reply(200, {
      banners: [],
    })
    const { asFragment } = render(
      <App>
        <Banners />
      </App>,
    )
    await waitFor(() => screen.findByTestId("test-banners"))
    expect(asFragment()).toMatchSnapshot()
    queryClient.clear()
  })

  it("render well with images", async () => {
    queryClient.clear()
    nock("https://81.69.200.140")
      .get(`/api/banner?type=2`)
      .reply(200, {
        banners: [
          { pic: "http://image-one" },
          { pic: "http://image-two" },
          { pic: "" },
        ],
      })
    const { asFragment } = render(
      <App>
        <Banners />
      </App>,
    )

    await waitFor(() => screen.findAllByRole("img"))
    expect(asFragment()).toMatchSnapshot()
    fireEvent.click(screen.getAllByRole("listitem")[0])
    queryClient.clear()
  })
})
