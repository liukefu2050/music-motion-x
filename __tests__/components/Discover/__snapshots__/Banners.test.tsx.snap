// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banners render well with images 1`] = `
<DocumentFragment>
   
  <div
    class=" w-full relative overflow-hidden"
    data-testid="test-banners"
    style="padding: 39% 0px 0px;"
  >
    <div
      class=" absolute bottom-1 left-1/2 -translate-x-1/2 z-20 will-change-transform"
    >
      <li
        class=" mr-1 w-[6px] h-[6px] inline-block rounded-[50%] bg-secondary"
      />
      <li
        class=" mr-1 w-[6px] h-[6px] inline-block rounded-[50%]  bg-white"
      />
      <li
        class=" mr-1 w-[6px] h-[6px] inline-block rounded-[50%]  bg-white"
      />
    </div>
    <div
      class=" absolute w-full top-0 left-0 whitespace-nowrap touch-pan-x"
    >
      <div
        class=" w-full inline-block transition-transform ease-in-out duration-700 scale-100 "
      >
        <img
          alt=""
          class=" w-full rounded-[10px]"
          draggable="false"
          src="https://image-one"
        />
      </div>
      <div
        class=" w-full inline-block transition-transform ease-in-out duration-700 scale-100 "
      >
        <img
          alt=""
          class=" bg-dg select-none  my_image  w-full rounded-[10px] pika-lazy"
          data-settled="false"
          data-src="https://image-two"
          draggable="false"
        />
      </div>
      <div
        class=" w-full inline-block transition-transform ease-in-out duration-700 scale-100 "
      >
        <img
          alt=""
          class=" bg-dg select-none  my_image  w-full rounded-[10px] "
          data-settled="false"
          data-src=""
          draggable="false"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Banners render well without images 1`] = `
<DocumentFragment>
   
  <div
    class=" w-full relative overflow-hidden"
    data-testid="test-banners"
    style="padding: 39% 0px 0px;"
  >
    <div
      class=" absolute bottom-1 left-1/2 -translate-x-1/2 z-20 will-change-transform"
    />
  </div>
</DocumentFragment>
`;
